src/libraries/System.Linq/src/System/Linq/Reverse.SpeedOpt.cs

```diff
@@ -7,28 +7,28 @@ namespace System.Linq
 {
     public static partial class Enumerable
     {
-        private sealed partial class ReverseIterator<TSource> : IPartition<TSource>
+        private sealed partial class ReverseIterator<TSource>
```

Comment:
Similar to the Range and Repeat iterators, couldn't this type also implement `IList<T>` and `IReadOnlyList<T>`? It might be possible to eliminate some of the boilerplate using an intermediate internal interface type that implements the throwing mutable APIs using DIM, i.e.
```C#
internal interface IListIterator<T> : IList<T>, IReadOnlyList<T>, IList
{
    void ICollection<T>.Add(T item) => throw new NotSupportedException();
    /* ... */
}

internal sealed class ReverseIterator<T> : IListIterator<T>
{
}
```

Reply:
> couldn't this type also implement IList<T> and IReadOnlyList<T>?

I don't think so? The ReverseIterator could be wrapping an arbitrary enumerable; it would mean for example that every indexing into the list could necessitate iterating through the enumerable, which could incur untold cost.